doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,112 @@
+tests verifying that the operator actually works are not possible to do in a generic way. This leads to some of the functional tests in the current scorecard
+implementation to be inaccurate or too insufficient to be useful. For more useful functional scorecard tests, we need to allow some basic user input for tests.
+
+## Goals
+
+- Implement user-defined scorecard tests
+- Replace existing "Operator actions are reflected in status" and "Writing into CRs has effects" tests with user-defined tests
+
+## Design overview
+
+A new section would be added to the scorecard config file called `functional_tests` where a user can define various aspects of a test. For example, this definition runs a similar test to the memcached-operator scale test from the SDK's e2e test:
+
+```yaml
+functional_tests:
+  - cr: "deploy/crds/memcached.cr.yaml"
+    expected_resources:
+      - kind: Deployment
+        name: "example_memcached"
+        fields:
+          status:
+            readyReplicas: 3
+          spec:
+            template:
+              spec:
+                containers:
+                  - image: memcached:1.4.36-alpine
+    expected_status:
+      scorecard_function_length:
+        nodes: 3
+    modifications:
+      - spec:
+        - size: 4
+        expected_resources:
+          - kind: Deployment
+            name: "example_memcached"
+            fields:
+              status:
+                readyReplicas: 4
+        expected_status:
+          scorecard_function_length:
+            nodes: 4
+
+```
+
+This is what the golang structs would look like, including comments describing each field:
+
+```go
+// Struct containing a user defined test. User passes tests as an array using the `functional_tests` viper config
+type UserDefinedTest struct {
+    // Path to cr to be used for testing
+    CRPath string `mapstructure:"cr"`
+    // Resources expected to be created after the operator reacts to the CR
+    ExpectedResources []ExpectedResource `mapstructure:"expected_resources"`
+    // Expected values in CR's status after the operator reacts to the CR
+    ExpectedStatus map[string]interface{} `mapstructure:"expected_status"`
+    // Sub-tests modifying a few fields with expected changes
+    Modifications []Modification `mapstructure:"modifications"`
+}
+
+// Struct containing a resource and its expected fields
+type ExpectedResource struct {
+    // Kind of resource
+    Kind string `mapstructure:"kind"`
+    // Name of resource
+    Name string `mapstructure:"name"`
+    // The fields we expect to see in this resource
+    Fields map[string]interface{} `mapstructure:"fields"`
+}
+
+// Modifications specifies a spec field to change in the CR with the expected results
+type Modification struct {
+    // a map of the spec fields to modify
+    Spec map[string]interface{} `mapstructure:"spec"`
```

Comment:
There are probably tests that involve changes outside of the `spec` (e.g. changing annotations). Should we account for that type of thing?

Also, defining a modification this way has some drawbacks, but I think it works for most use cases.

A few examples of when it doesn't work well are for modifying slices and removing keys. When a modified value is a slice, the entire modified slice has to be included in the modification. Also, it isn't possible to remove a key, only set it to `nil`. Removing a key might be necessary when there's a semantic difference between not present and `nil`.

For non-simple use cases, we may want an alternative ways to specify modifications. Perhaps the YAML-equivalent of [JSON patch](https://tools.ietf.org/html/rfc6902). An implementation for YAML exists (https://github.com/krishicks/yaml-patch) if that's something we want to look into.

Somewhat tangentially, another benefit of JSON patch is that there's a `test` operation that may be useful for implementing the checks.

Reply:
Using a JSON patch for the modification sounds like a good idea. It's a standard and is quite simple to understand.

Do you think it would make sense to add a JSON patch var to the modification struct and support both using `Spec` and JSON patches, or do you think we should only support 1 method for simplicity?

Also, do you you think it would be a good idea to add JSON patch `test` operation support to the `expected_resource` and `expected_status` types? That could make it easier for some test cases, especially ones involving arrays, where the current proposal would need to do something like this to get an item in the second element of an array:

```yaml
spec:
  containers:
    -
    - image: my-image
```

vs

```yaml
json_patch: |
  {"op":"test","path":"/spec/containers/1/image","value":"my-image"}
```

Reply:
+1 for supporting JSON patch fields to specify both modifications and the expected resources/status.

Reply:
I don't think the map-style declaration would cause many issues when it comes to how we compare things. There are only 3 non-map/non-array values that you can have in go-yaml/json: number (which can be int/int64/float64; we can just convert any of these types to float64 before comparison), string, and bool. When it comes to maps and array, we just walk down them, not really any different than a JSONPatch would. The main benefit of JSONPatch is that we can more easily define array indices and we can also delete values.

The main benefit of keeping a map-style declaration is that we can have the scorecard functions like the array length checker (another potential example would be a regex matcher). We can't implement that with JSONPatch. We also get to keep the same structure as the actual kubernetes resources, which may be a bit clearer for users.